src/repo/fs.rs

```diff
@@ -5,147 +5,381 @@ use async_trait::async_trait;
-        Ok(Some(block))
+
+        let path = block_path(self.path.clone(), &cid);
+
+        // probably best to do everything in the blocking thread if we are to issue multiple
+        // syscalls
+        tokio::task::spawn_blocking(|| {
+            let mut file = match std::fs::File::open(path) {
+                Ok(file) => file,
+                Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(None),
+                Err(e) => {
+                    return Err(e.into());
+                }
+            };
+
+            let len = file.metadata()?.len();
+
+            let mut data = Vec::with_capacity(len as usize);
+            file.read_to_end(&mut data)?;
+            let block = Block::new(data.into_boxed_slice(), cid);
+            Ok(Some(block))
+        })
+        .await?
     }
 
     async fn put(&self, block: Block) -> Result<(Cid, BlockPut), Error> {
-        let path = block_path(self.path.clone(), &block.cid());
-        let cids = &self.cids;
-        let data = block.data();
-        let mut file = fs::File::create(path).await?;
-        file.write_all(&*data).await?;
-        file.flush().await?;
-        let retval = if cids.lock().await.insert(RepoCid(block.cid().to_owned())) {
-            BlockPut::NewBlock
-        } else {
-            BlockPut::Existed
+        use std::collections::hash_map::Entry;
+
+        let target_path = block_path(self.path.clone(), &block.cid());
+        let cid = block.cid;
+        let data = block.data;
+
+        // why synchronize here? because when we lose the race we cant know if there was someone
+        // else interested in writing this block or not
+
+        let (tx, mut rx, created) = {
+            let mut g = self.writes.lock().expect("cant support poisoned");
+
+            match g.entry(RepoCid(cid.to_owned())) {
+                Entry::Occupied(oe) => {
+                    // someone is already writing this, nice
+                    (oe.get().clone(), oe.get().subscribe(), false)
+                }
+                Entry::Vacant(ve) => {
+                    // we might be the first, or then the block exists already
+                    let (tx, rx) = broadcast::channel(1);
+                    ve.insert(tx.clone());
+                    (tx, rx, true)
+                }
+            }
         };
-        // FIXME: checking if the file existed already while creating complicates this function a
-        // lot; might be better to just guard with mutex to enforce single task file access.. the
-        // current implementation will write over the same file multiple times, each time believing
-        // it was the first.
-        Ok((block.cid().to_owned(), retval))
+
+        fs::create_dir_all(
+            target_path
+                .parent()
+                .expect("parent always exists as it's the sharding directory"),
+        )
+        .await?;
+
+        // pick the next writer by chance
+        match std::fs::OpenOptions::new()
```

Comment:
yeah forgot to change this back to async one; I'll probably throw it in the spawn_blocking one.